scalameta/dialects/shared/src/main/scala/scala/meta/Dialect.scala

```diff
@@ -0,0 +1,427 @@
+    val allowViewBounds: Boolean,
+    // Are `with` intersection types supported by this dialect?
+    val allowWithTypes: Boolean,
+    // Are XML literals supported by this dialect?
+    // We plan to deprecate XML literal syntax, and some dialects
+    // might go ahead and drop support completely.
+    val allowXmlLiterals: Boolean,
+    // What kind of separator is necessary to split top-level statements?
+    // Normally none is required, but scripts may have their own rules.
+    val toplevelSeparator: String,
+    // Are numeric literal underscore separators, i.e. `1_000_000` legal or not?
+    val allowNumericLiteralUnderscoreSeparators: Boolean
+) extends Product
+    with Serializable {
+
+  // NOTE(olafur) checklist for adding a new dialect field in a binary compatible way:
+  // - add new field to primary constructor.
+  // - add new parameter to `privateCopy()` method.
+  // - add new `withX()` method to allow overriding that specific field.
+  // - add new `x = FIELD_DEFAULT_VALUE` in the `def this()` constructor below.
+
+  def this(
+      allowAndTypes: Boolean,
+      allowAtForExtractorVarargs: Boolean,
+      allowCaseClassWithoutParameterList: Boolean,
+      allowColonForExtractorVarargs: Boolean,
+      allowEnums: Boolean,
+      allowImplicitByNameParameters: Boolean,
+      allowImplicitFunctionTypes: Boolean,
+      allowInlineIdents: Boolean,
+      allowInlineMods: Boolean,
+      allowLiteralTypes: Boolean,
+      allowMethodTypes: Boolean,
+      allowMultilinePrograms: Boolean,
+      allowOrTypes: Boolean,
+      allowPatUnquotes: Boolean,
+      allowSpliceUnderscores: Boolean,
+      allowTermUnquotes: Boolean,
+      allowToplevelTerms: Boolean,
+      allowTrailingCommas: Boolean,
+      allowTraitParameters: Boolean,
+      allowTypeLambdas: Boolean,
+      allowViewBounds: Boolean,
+      allowWithTypes: Boolean,
+      allowXmlLiterals: Boolean,
+      toplevelSeparator: String
+  ) = {
+    this(
+      allowAndTypes,
+      allowAtForExtractorVarargs,
+      allowCaseClassWithoutParameterList,
+      allowColonForExtractorVarargs,
+      allowEnums,
+      allowImplicitByNameParameters,
+      allowImplicitFunctionTypes,
+      allowInlineIdents,
+      allowInlineMods,
+      allowLiteralTypes,
+      allowMethodTypes,
+      allowMultilinePrograms,
+      allowOrTypes,
+      allowPatUnquotes,
+      allowSpliceUnderscores,
+      allowTermUnquotes,
+      allowToplevelTerms,
+      allowTrailingCommas,
+      allowTraitParameters,
+      allowTypeLambdas,
+      allowViewBounds,
+      allowWithTypes,
+      allowXmlLiterals,
+      toplevelSeparator,
+      allowNumericLiteralUnderscoreSeparators = false
+      // NOTE(olafur): declare the default value for new fields above this comment.
+    )
+  }
+
+  // Are unquotes ($x) and splices (..$xs, ...$xss) allowed?
+  def allowUnquotes: Boolean = allowTermUnquotes || allowPatUnquotes
+
+  def withAllowAndTypes(allowAndTypes: Boolean): Dialect = {
```

Comment:
optional: why not call the parameter as simply `flag` or even `f`, `b` (for bool), `v` (for value), to avoid repetition/clutter and to shorten (although i remember that you prefer `unfold`)? there's clear semantic meaning.
```
  def withAllowAndTypes(v: Boolean): Dialect = privateCopy(allowAndTypes = v)
```

Reply:
Good suggestion. I renamed the parameter to `newValue` so that it's easier to copy-paste new methods.